other/lfu_cache.py

```diff
@@ -0,0 +1,188 @@
+from typing import Callable, Optional
+
+
+class DoubleLinkedListNode:
+    '''
+    Double Linked List Node built specifically for LFU Cache
+    '''
+
+    def __init__(self, key: int, val: int):
+        self.key = key
+        self.val = val
+        self.freq = 0
+        self.next = None
+        self.prev = None
+
+
+class DoubleLinkedList:
+    '''
+    Double Linked List built specifically for LFU Cache
+    '''
+
+    def __init__(self):
+        self.head = DoubleLinkedListNode(None, None)
+        self.rear = DoubleLinkedListNode(None, None)
+        self.head.next, self.rear.prev = self.rear, self.head
+
+    def add(self, node: DoubleLinkedListNode) -> None:
+        '''
+        Adds the given node at the head of the list and shifting it to proper position
+        '''
+
+        temp = self.rear.prev
+
+        self.rear.prev, node.next = node, self.rear
+        temp.next, node.prev = node, temp
+        node.freq += 1
+        self._position_node(node)
+
+    def _position_node(self, node):
+        while node.prev.key and node.prev.freq > node.freq:
+            node1, node2 = node, node.prev
+
+            node1.prev, node2.next = node2.prev, node1.prev
+            node1.next, node2.prev = node2, node1
+
+    def remove(self, node: DoubleLinkedListNode) -> DoubleLinkedListNode:
+        '''
+        Removes and returns the given node from the list
+        '''
+
+        temp_last, temp_next = node.prev, node.next
+        node.prev, node.next = None, None
+        temp_last.next, temp_next.prev = temp_next, temp_last
+
+        return node
+
+
+class LFUCache:
+    '''
+    LFU Cache to store a given capacity of data. Can be used as a stand-alone object
+    or as a function decorator.
+
+    >>> cache = LFUCache(2)
```

Comment:
Could `LFUCache` take optional cache parameter? Would be easier testable.

```
 def __init__(self, capacity: int, cache: dict = None):
        self.list = DoubleLinkedList()
        self.capacity = capacity
        self.num_keys = 0
        self.hits = 0
        self.miss = 0
        self.cache = cache or {}
```

Reply:
That seems like a big change.